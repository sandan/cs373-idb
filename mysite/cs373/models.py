
from django.db import models


class Stage(models.Model):
    """
    A reprentation of a stage
    """
    name            = models.CharField(max_length=400, unique=True)
    #sponsor        = models.ForeignKey(Sponsor)

    # def get_name(self):
    #     return self.name

    def get_url(self):
        """
        returns url
        """
        return "/stages/%s/" % self.id

    def __str__(self):
        """
        returns stage name
        """
        return self.name

class Sponsor(models.Model):
    """
    A representation of a Sponsor
    """

    name            = models.CharField(max_length=400)
    business_type   = models.CharField(max_length=400)
    stage           = models.ForeignKey(Stage, blank=True, null=True)


    # def get_name(self):
    #     return self.name

    # def get_business_type(self):
    #     return self.busoness_type

    def get_url(self):
        """
        returns url
        """
        return "/sponsors/%s/" % self.id

    def img_url(self):
        return '/images/sponsor/%i.jpg' % self.name.lower().replace(' ','')

    def __str__(self):
        """
        returns Sponsor name
        """
        return self.name


class Artist(models.Model):
    """
    A representation of an artist
    """
    name            = models.CharField(max_length=400, unique=True)
    label           = models.CharField(max_length=400)
    origin          = models.CharField(max_length=400)
    genre           = models.CharField(max_length=400)
    stage           = models.ForeignKey(Stage)

    def get_url(self):
        """
        returns url
        """
        return "/artists/%s/" % self.id

   # def photo(self):
    #    return '/static/images/artist/%s.jpg' % self.name.replace(' ','').lower()

    def __str__(self):
        """
        returns Artist name
        """
        return self.name

class Media(models.Model):
    """
    Media resource for Artist, Sponsor, Stage
    """
    
    Bio=models.CharField(max_length=5000)
    Photo= models.CharField(max_length=400)
    Youtube=models.CharField(max_length=400)
    Video=models.CharField(max_length=400)
    Twitter=models.CharField(max_length=400)
    Facebook=models.CharField(max_length=400)
    # holds the html code for twitter timeline widget
    twitterwidget = models.CharField(max_length=400)
    # holds the src of youtube video
    # ex. src='{{media.youtubevideo}}'
    youtubevideo = models.CharField(max_length=400)
    Webpage=models.CharField(max_length=400)
    
    def __str__(self):
        """
        returns Webpage link
        """
        return self.Webpage
        
class ArtistMedia(Media):
    ar = models.ForeignKey(Artist)

class StageMedia(Media):
    st = models.ForeignKey(Stage)
    
class SponsorMedia(Media):
    sp = models.ForeignKey(Sponsor)
    
"""
DEPRECATED
class Member(models.Model):
    
    
    
    first_name      = models.CharField(max_length=400)
    last_name       = models.CharField(max_length=400)
    artist          = models.ForeignKey(Artist)

    def __str__(self):
    
    
    
        ret_str = self.first_name
        if(self.last_name):
            ret_str += " " + self.last_name
        return ret_str


"""