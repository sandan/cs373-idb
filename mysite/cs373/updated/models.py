from django.db import models

"""
Three Main Entities
..that we are modeling the Austin City Limits Festival with
"""

class Stage(models.Model):
    """
    Stages are physical locations on the acl festival map:
    see theaustinites.pythonanywhere.com/stages/

    Locations are mappings from the physical stage to an integer.
    For any single year, a stage will have one sponsor and many artists playing.
    For different years, a stage can potentially have different sponsors and many artists playing.
    """
    location            = models.PositiveSmallIntegerField(unique=True)
    def __str__(self):
        """
        returns location
        """
        return str(self.location)

class Sponsor(models.Model):
    """
    A Sponsor is an entity that purchases a stage
    for Music Artists to play on.
    For any single year, a sponsor can only sponsor one stage.
    For many different years a sponsor could potentially sponsor
    a different stage (location).
    """
    name            = models.CharField(max_length=255, unique=True)
    industry            = models.CharField(max_length=255)

    def __str__(self):
        """
        returns name
        """
        return self.name

class Artist(models.Model):
    """
    An Artist is an entity that plays on a sponsored stage.
    For any single year, an artist can only play on one stage.
    For multiple different years, an artist may play on potentially >1 stage.
    """
    name             = models.CharField(max_length=255, unique=True)
    label            = models.CharField(max_length=255)
    genre            = models.CharField(max_length=255)
    origin           = models.CharField(max_length=255)

    def __str__(self):
        """
        returns name
        """
        return self.name

"""
TIME
relationship classes
"""
# https://docs.djangoproject.com/en/1.6/ref/models/instances/#django.db.models.Model

class stage_sponsor_yr(models.Model):
    """
    create with method not the constructor
    """
    stage           = models.ForeignKey(Stage)
    sponsor         = models.ForeignKey(Sponsor)
    date            = models.DateField()      #datetime
    key             = models.CharField(max_length=255, unique=True)


    @classmethod
    def create(self, stage, sponsor, date):
        """
        For initialization of the primary key, Django doesn't support multi-column pk's.
        This is needed to enforce the data integrity between sponsors and stages in relation to time.
        """
        assert type(stage) == Stage
        assert type(sponsor) == Sponsor

        pkey=str(sponsor.id)+str(date.year)
        instance=self(stage=stage, sponsor=sponsor, date=date, key=pkey)
        return instance

    def get_yr(self):
        return self.date.year

class stage_artist_yr(models.Model):
    """
    For initialization of the primary key, Django doesn't support multi-column pk's.
    This is needed to enforce the data integrity between artists and stages in relation to time.
    """
    stage           = models.ForeignKey(Stage)
    artist          = models.ForeignKey(Artist)
    date            = models.DateField()      #datetime
    key             = models.CharField(max_length=255, unique=True)

    @classmethod
    def create(self, stage, artist, date):
        """
        For initialization of the primary key, Django doesn't support multi-column pk's.
        This is needed to enforce the data integrity between sponsors and stages in relation to time.
        ex: relationship=stage_artist_yr.create(stage, artist, (datetime) yr)
        """

        assert type(stage) == Stage
        assert type(artist) == Artist

        pkey=str(artist.id)+str(date.year)
        instance=self(stage=stage, artist=artist, date=date,key=pkey)
        return instance

    def get_yr(self):
        return self.date.year

"""
MEDIA
info for dynamic webpages
"""
class Media(models.Model):
    """
    Media resource for Artist, Sponsor, Stage
    """

    bio             = models.TextField()
    photo           = models.URLField(max_length=255)
    youtube         = models.URLField(max_length=255)
    video           = models.URLField(max_length=255)
    youtubevideo    = models.URLField(max_length=255)
    twitter         = models.URLField(max_length=255)
    twitterwidget   = models.URLField(max_length=255)
    facebook        = models.URLField(max_length=255)
    webpage         = models.URLField(max_length=255)

    def __str__(self):
        return self.webpage

class ArtistMedia(Media):
    artist           = models.ForeignKey(Artist, unique=True)


class StageMedia(Media):
    name            = models.CharField(max_length=42) #Derivable
    year            = models.DateField()           #datetime
    stage           = models.ForeignKey(Stage)


class SponsorMedia(Media):
    sponsor      = models.ForeignKey(Sponsor, unique=True)


