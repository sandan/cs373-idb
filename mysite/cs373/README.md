cs373-idb
=========

Group Name:       The Austinites
Group Members:    Stephen Ridings, Mark Sandan, Jesus Hernandez, Carlos Rodriguez, Sheeyla Garcia, Kyle Nicola
Proposed project: Austin City Limits (ACL)
Example Pages:    Artists, Sponsors and Events



There will be three phases.

Phase I Features:
-   Collect data on ten artists, ten sponsors, and all the events (7).
    Identify common attributes of the data and insure that every data item has a value for that attribute.
    Make sure to collect enough information to make it a rich experience.
-   Create a RESTful API using Apiary. Log into Apiary with your GitHub credentials and connect Apiary to your private repos.
    Think carefully about the types and format of the requests and responses, primary keys, foreign key constraints, and validations.
-   Create a set of Django models to represent the data and that corresponds to the API.
    Think carefully about the types of the models and about the many-to-many relationships between artists, sponsors, and events.
-   Create a set of unit tests of the Django models.
-   Create a static HTML page for three unique artists, three unique sponsors, and three unique events that displays all of the data collected.
    The pages must be served by Django on PythonAnywhere using Django templates to embed the static content into HTML.
    Design the splash page well. Design the navigation well.
    Use Twitter Bootstrap to present information in an organized way. Make sure to embed the different media. Make sure that everything that can be linkable is.
    Create a responsive design that looks good on mobile devices.
-   Write an initial technical report.
    The audience comprises other software developers, as opposed to users.
    For this phase make it at least 10 pages.
    Document the RESTful API and the Django models well.
    Format the report clearly, attractively, and consistently, using good headers, figures, and grammar.



The pages MUST contain the following:

Splash Page Features:
- the group name
- the group members
- a way to navigate to the nine static HTML pages

Artist Page Features:
- name, members, genre, label(s), origin, date and time, citations, external links
* small artist description (iTunes, Spotify, Wikipedia, etc.)
- embedded images (e.g. Bing, Flickr, Google)
- embedded videos (e.g. Youtube, Google, Vimeo)
- embedded maps (e.g. Google Maps)
- embedded social network feeds (e.g. Facebook, Twitter)
- sponsors
- events

Sponsor Page Features:
- name, business type, location, history, CEO, contact info, citations, external links
- embedded images, logo (e.g. Bing, Flickr, Google)
- embedded videos, ad (e.g. Youtube, Google, Vimeo)
- embedded maps, HQ (e.g. Google Maps)
- embedded social network feeds (e.g. Facebook, Twitter)
- artists
- events

Event Page Features:
- name, kind, location, citations, external links
- embedded images (e.g. Bing, Flickr, Google)
- embedded videos (e.g. Youtube, Google, Vimeo)
- embedded maps (e.g. Google Maps, location during festival)
- embedded social network feeds (e.g. Facebook, Twitter)
- artists
- sponsors



The technical report MUST contain the following:
- Title:        The group name, the group members.
- Introduction: What is the problem? What are the use cases?
- Design:       RESTful API, Django models.
- Tests:        unit tests of the Django models.
- Other:        Proof-read your report. Get another group to read it. Read it aloud. 
                Create diagrams with captions.
                Create sections and subsections effectively.



Groups:
- Pick a project leader for the group. A group member can only be project leader once.
- Once the grader has graded the project, that will be the project grade.
- The project leader will always receive the project grade.
- The project leader can add or subtract up to 10 points to or from any member of the group in exchange for
  adding or subtracting the same amount of points from the other member(s) of the group.
- For example, the leader could add 10 points to person A, and subtract 5 points from person B and person C.
- The leader will report those adjustments in the Google Form.



Requirements:
-   Estimate time to completion.
-   Create a public Git repository at GitHub, named cs373-idb.
-   Add these requirements to the issue tracker at GitHub, at least 10 issues.
    Add at least 10 more issues, one for each bug or feature, both open and closed with a good description and a label.
-   Clone your public code repo onto your local directory.
-   Make at least 5 commits, one for each bug or feature.
    If you cannot describe your changes in a sentence, you are not committing often enough.
    Make meaningful commit messages identifying the corresponding issue in the issue tracker (see here).
-   Write unit tests in tests.py that test corner cases and failure cases until you have an average of 3 tests for each Django model     call, confirm the expected failures, and add, commit, and push to the public code repo.
-   Implement and debug the simplest possible solution with assertions that check pre-conditions, post-conditions, argument validity,     and return-value validity, until all tests pass, and add, commit, and push to the private code repo.
-   Run pydoc on models.py, which will create models.html, that then documents the interfaces to your models.
    Create inline comments if you need to explain the why of a particular implementation.
    Use a consistent coding convention with good variable names, good indentation, blank lines, and blank spaces.
-   Create a log of your commits in IDB.log.
-   Obtain the git SHA with git rev-parse HEAD
-   Fill in the Google Form.
-   It is your responsibility to protect your code from the rest of the students in the class. If your code gets out, you are as         guilty as the recipient of academic dishonesty.



Requirements for getting a non-zero grade.
[ 5 pts] GitHub public repo with a log of the commits.
[ 5 pts] GitHub issue tracker with issues from requirements and more.
[10 pts] RESTful API at Apiary.
[10 pts] Set of Django models.
[10 pts] Website using Django and Twitter Bootstrap at PythonAnywhere.
[10 pts] Average of 3 unit tests per Django model call.
[ 5 pts] Pydoc documentation.
[10 pts] Technical Report, double-spaced, 10 pages.
[ 5 pts] Google Form with time estimate.



Files:
  apiary.apib
  models.html
  Report.pdf
  tests.out
  tests.py
