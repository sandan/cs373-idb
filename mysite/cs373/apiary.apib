FORMAT: 1A
HOST: http://theaustinites.pythonanywhere.com/

# Austin City Limits Project API
ACL Project is a representation of the bands, stages, events, and artists appearing at 
**Austin City Limits**.

The API allows **GET** requests to the following:

* Stages
* Sponsors
* Artists
* ArtistMedia
* StageMedia
* SponsorMedia

**Note:** The ACL API allows information to be selected on a per year basis. 
The earliest available year starts at 2012.

# Group Stage
Stage related resources of the Austin City Limits Project API

## Stage [/api/stages/{location}/]
A single stage object.   

The Stage resource has the following attributes:

- location

The stage's *location* is assigned by the ACL Project API at the moment of creation.
The Stage resource represents a physical location on the Austin City Limits map.
see http://theaustinites.pythonanywhere.com/stages/.
Note that depending on the year of the festival, a stage may have a different
sponsor sponsoring that stage or different artists playing on that stage.

+ Parameters
    + location (integer) ...  a number
    
+ Model (application/json)
    JSON representation of Stage resource.

    + Body
    
            {
                "years": [2011,2012,2013,2014]
            }

### Retrieve a single Stage [GET]
A GET request returns a json dictionary with a key "year"
and the years that that particular stage location has been in use for the 
festival.

+ Response 200 (application/json)

    [Stage][]
        
## Stage Collection [/api/stages/{location}/{year}]
A stage collection returns information about it's *sponsor*
and *artists* that play on that stage for the given year. 

+ Model (application/json)

        JSON representation of the Stage Collection Resource.  The Stage resources in the collections are embedded.

        + Body
            
            [
                {   "location": 1
                    "sponsor": 'Honda',
                    "artists": ['Kendrick Lamar','Skrilex',...]
                }
            ]

### List Stage Information for a given year [GET]
A GET request for a **Stage Collection** returns the *location*, *sponsor* for 
the requested stage location and year as well as a *list of artists*
that have played on that stage during the requested year.
+ Response 200
    
    [Stage Collection][]
    
# Group Sponsor
Sponsor related resources of the **ACL API**

## Sponsor [/api/sponsors/{id}/]
A single sponsor object.

The Sponsor resource has the following attributes:
- id
- name
- industry

+ Parameters
    + id (integer) ... ID of the Sponsor in the form of a number
    
+ Model (application/json)
    JSON representation of Sponsor resource.

    + Body
    
            {
                "id": 42,
                "name": "Tito's Handmade Vodka",
                "industry": "alcohol industry"
            }

### Retrieve a single Sponsor [GET]
A GET request for a sponsor with given id returns 
the *id*, *name*, and *industry* the sponsor is associated with.
Note that sponsor names are unique.
+ Response 200 (application/json)

    [Sponsor][]


## Sponsor Collection [/api/sponsors/year/{year}]
Collection of all Sponsors that sponsored a Stage for the requested year.
+ Model (application/json)

        JSON representation of the Sponsor Collection Resource.  
        The Sponsor resources in the collections are embedded.

        + Body
            
            [
                {
                    "sponsors": ['id of sponsor 1','id of sponsor 2', etc...]
            
                }
            ]

            
### List All Sponsors for a given year[GET]
A GET request returns a json dictionary with a sponsors key and a list
of ids corresponding to a sponsor.

+ Response 200
    
    [Sponsor Collection][]

    
# Group Artist
Artist related resources of the **ACL API**

## Artist [/api/artists/{id}/]
A single Artist object with all its details

The Artist resource has the following attributes:

- id
- name
- label
- origin
- genre

The artist **id** is assigned by the ACL Project API at the moment of creation.

+ Parameters
    + id (required, number, `1`) ... Numeric `id` of the Artist to perform action with.

    
+ Model (application/json)
    JSON representation of Artist resource.

    + Body
    
            [   
                {
            
                "id": 42,
                "name": "Artist name"
                "label": "Label of artist"
                "origin": "Where the artist was from"
                "genre": "Genre of the artist"
                
                }
            ]   


### Retrieve an Artist [GET]
+ Response 200 (application/json)

    [Artist][]
    

## Artist Collection [/api/artists/year/{year}]
Collection of all Artists for a given year

It **embeds** *Artist Resources* in the ACL Project API

+ Model (application/json)

        JSON representation of the Sponsor Collection Resource.  The Sponsor resources in the collections are embedded.

        + Body
            [
                {
                    "artists": [id of artist,...]
                    
                }
            ]

            
### List All Artists [GET]
A GET request returns a json dictionary with an artists key whose value
is a list of artist id's.
+ Response 200
    
    [Artist Collection][]

# Group Media
Media related resources for Artist, Stage, and Sponsor

## ArtistMedia [/api/artists/{id}/media/]
A single ArtistMedia object with all its details

The ArtistMedia resource has the following attributes:

- ar (the artist)
- bio
- photo
- youtube
- video
- twitter
- facebook
- twitterwidget
- youtubevideo
- webpage

+ Parameters
    + id (required, number, `1`) ... Numeric `id` of the ArtistMedia to perform action with.

    
+ Model (application/json)
    JSON representation of ArtistMedia resource.

    + Body
    
            {
             "ar" : 42
             "bio" : "a description"
             "photo" : "url"
             "youtube"  "url"
             "video" : "url"
             "twitter" : "url"
             "facebook" : "url"
             "twitterwidget" : "twitter handle"
             "webpage" : "url"
            }
            
### Retrieve an ArtistMedia [GET]
+ Response 200 (application/json)

    [ArtistMedia][]

## SponsorMedia [/api/sponsors/{id}/media/]
A single SponsorMedia object with all its details

The SponsorMedia resource has the following attributes:

- sp (the sponsor)
- bio
- photo
- youtube
- video
- twitter
- facebook
- twitterwidget
- youtubevideo
- webpage

+ Parameters
    + id (required, number, `1`) ... Numeric `id` of the SponsorMedia to perform action with.

    
+ Model (application/json)
    JSON representation of ArtistMedia resource.

    + Body
    
            {
             "sp" : 42
             "bio" : "a description"
             "photo" : "url"
             "youtube"  "url"
             "video" : "url"
             "twitter" : "url"
             "facebook" : "url"
             "twitterwidget" : "twitter handle"
             "webpage" : "url"
            }
            
### Retrieve an SponsorMedia [GET]
+ Response 200 (application/json)

    [SponsorMedia][]
    
## StageMedia [/api/stages/{id}/media/]
A single StageMedia object with all its details

The StageMedia resource has the following attributes:

- st (the stage)
- bio
- photo
- youtube
- video
- twitter
- facebook
- twitterwidget
- youtubevideo
- webpage

+ Parameters
    + id (required, number, `1`) ... Numeric `id` of the StageMedia to perform action with.

    
+ Model (application/json)
    JSON representation of StageMedia resource.

    + Body
    
            {
             "st" : 42
             "bio" : "a description"
             "photo" : "url"
             "youtube"  "url"
             "video" : "url"
             "twitter" : "url"
             "facebook" : "url"
             "twitterwidget" : "twitter handle"
             "webpage" : "url"
            }
            
### Retrieve an StageMedia [GET]
+ Response 200 (application/json)

    [StageMedia][]
