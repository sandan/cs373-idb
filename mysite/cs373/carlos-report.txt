\section{Updates}
Updates for the third iteration at a glance:
\begin{itemize}
 \item An About page has been created to provide an overview of our website
\end{itemize}
\begin{itemize}
 \item Splash page and web design have been tweaked to include better navigation and usability
\end{itemize}
\begin{itemize}
 \item Search capability implemented using Haystack and Whoosh
\end{itemize}
\begin{itemize}
 \item Created a dynamic page implementing another group's API in order to test it 
\end{itemize}
\begin{itemize}
 \item Created a presentation that overviews all three phases of our concluding IDB project
\end{itemize}

Summary
The main changes that have been made can be summarized into three main categories: search implementation, model refactoring, and website layout tweakings.
In order to implement the search functionality we installed Haystack to utilize as the search framework and Whoosh as the search engine.
The models were drastically changed to include intermediary models. These models represented relationships and represent the concept of time, which would allow us to expand our database to include artists, sponsors and stages used over the years.
The website itself was also expanded to include an About page with information about our team and the Splash page was also updated.

Technology Stack
The technologies used are:
• PythonAnywhere: a web hosting service with python environments supported.
Currently we use Python 3.4 and Django 1.6+.
• Twitter Bootstrap 3.2: a web hosting service with python environments supported Python 3.4, Django 1.6.
• Apiary: an online service to provide an API for client-side web access to our databases.
• MySQL: The current database backend using the mysql.connector.django engine.
• Haystack: A simplified search framework that allows you to use third-party search engines as its core.
• Whoosh: a full-text indexing and searching library implemented in pure Python.
