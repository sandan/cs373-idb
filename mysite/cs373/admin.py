from django.contrib import admin
from cs373.models import *
# Register your models here.

class StageMediaInline(admin.StackedInline):
    model = StageMedia
    extra = 0

class StageAdmin(admin.ModelAdmin):
    inlines = [StageMediaInline]

class SponsorMediaInline(admin.StackedInline):
    model = SponsorMedia
    extra = 0

class SponsorAdmin(admin.ModelAdmin):
    inlines = [SponsorMediaInline]

class ArtistMediaInline(admin.StackedInline):
    model = ArtistMedia
    extra = 0

class ArtistAdmin(admin.ModelAdmin):
    inlines = [ArtistMediaInline]


admin.site.register(Stage, StageAdmin)
admin.site.register(Sponsor, SponsorAdmin)
admin.site.register(Artist, ArtistAdmin)
