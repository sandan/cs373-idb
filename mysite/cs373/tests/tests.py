from django.test import TestCase
from cs373.models import *
from django.db.utils import IntegrityError
import datetime


class TypeTest(TestCase):

    #test the types of models created
    def test_types(self):
        a=Artist()
        s=Stage()
        p=Sponsor()

        self.assertEqual(type(a), Artist)
        self.assertEqual(type(p), Sponsor)
        self.assertEqual(type(s), Stage)

    #test types of relationship classes
    def test_types_relationships1(self):
        a=Artist()
        s=Stage()
        p=Sponsor()

        #correct artist type but not correct stage type
        try:
            stage_artist_yr.create(stage=p,artist=a,date=datetime.date(1999,3,3))
        except AssertionError:
            pass
        except Exception:
            self.fail("Unknown exception thrown!")
        else:
            self.fail("Expected exception to be thrown but it wasn't !")

        #correct stage type but not correct artist type
        try:
            stage_artist_yr.create(stage=s,artist=p,date=datetime.date(1999,3,3))
        except AssertionError:
            pass
        except Exception:
            self.fail("Unknown exception thrown!")
        else:
            self.fail("Expected exception to be thrown but it wasn't !")

        #both wrong types
        try:
            stage_artist_yr.create(stage=p,artist=a,date=datetime.date(1999,3,3))
        except AssertionError:
            pass
        except Exception:
            self.fail("Unknown exception thrown!")
        else:
            self.fail("Expected exception to be thrown but it wasn't !")

    def test_types_relationships2(self):
        a=Artist()
        s=Stage()
        p=Sponsor()

        try:
            stage_sponsor_yr.create(stage=s,sponsor=a,date=datetime.date(2100,3,3))
        except AssertionError:
            pass
        except Exception:
            self.fail("Unknown exception thrown!")
        else:
            self.fail("Expected exception to be thrown but it wasn't !")

        try:
            stage_sponsor_yr.create(stage=a,sponsor=p,date=datetime.date(1999,3,3))
        except AssertionError:
            pass
        except Exception:
            self.fail("Unknown exception thrown!")
        else:
            self.fail("Expected exception to be thrown but it wasn't !")

        try:
            stage_sponsor_yr.create(stage=a,sponsor=s,date=datetime.date(1999,3,3))
        except AssertionError:
            pass
        except Exception:
            self.fail("Unknown exception thrown!")
        else:
            self.fail("Expected exception to be thrown but it wasn't !")

class StageRelationshipTests(TestCase):

    #test creation and insertion
    def test_relationship1(self):
        s=Stage(location=2)
        s.save()

        a=Artist(name="hello")
        a.save()

        rel=stage_artist_yr.create(stage=s,artist=a,date=datetime.date(2005,10,2))
        rel.save()

        self.assertEqual(rel.get_yr(),2005)
        self.assertEqual(rel.key,str(a.id)+str(rel.get_yr()))
        self.assertEqual('hello',rel.artist.name)
        self.assertEqual(2,rel.stage.location)


    def test_relationship2(self):
        s=Stage(location=2)
        s.save()

        a=Sponsor(name="hello")
        a.save()

        rel=stage_sponsor_yr.create(stage=s,sponsor=a,date=datetime.date(2099,10,2))
        rel.save()

        self.assertEqual(rel.get_yr(),2099)
        self.assertEqual(rel.key,str(a.id)+str(rel.get_yr()))
        self.assertEqual('hello',rel.sponsor.name)
        self.assertEqual(2,rel.stage.location)

class ConstraintTests(TestCase):

    #same artist playing on two different stages of the same year
    def test_constraint_violation1(self):
        s=Stage(location=2)
        t=Stage(location=3)
        s.save()
        t.save()

        a=Artist(name='hello world')
        a.save()

        rel1=stage_artist_yr.create(stage=s,artist=a,date=datetime.date(2011,1,1))
        rel1.save()
        try:
            rel2=stage_artist_yr.create(stage=t,artist=a,date=datetime.date(2011,2,3))
            rel2.save()
        except IntegrityError:
            pass
        except Exception:
            self.fail("Unexpected Exception thrown!")
        else:
            self.fail("An exception expected to be thrown but wasn't!")

    #same sponsor sponsoring more than one stage in the same year
    def test_constraint_violation2(self):
        s=Stage(location=2)
        t=Stage(location=3)
        s.save()
        t.save()

        a=Sponsor(name='hello world')
        a.save()

        rel1=stage_sponsor_yr.create(stage=s,sponsor=a,date=datetime.date(2099,1,1))
        rel1.save()
        try:
            rel2=stage_sponsor_yr.create(stage=t,sponsor=a,date=datetime.date(2099,2,3))
            rel2.save()
        except IntegrityError:
            pass
        except Exception:
            self.fail("Unexpected Exception thrown!")
        else:
            self.fail("An exception expected to be thrown but wasn't!")

    #should be ok to save relationships on same stage if both are from a different year
    def test_no_constraint_violation1(self):
        s=Stage(location=2)
        t=Stage(location=3)
        s.save()
        t.save()

        a=Artist(name='hello world')
        a.save()

        rel1=stage_artist_yr.create(stage=s,artist=a,date=datetime.date(2012,1,1))
        rel1.save()

        rel2=stage_artist_yr.create(stage=s,artist=a,date=datetime.date(2011,1,1))
        rel2.save()


    #same deal for sponsors
    def test_no_constraint_violation2(self):
        s=Stage(location=2)
        t=Stage(location=3)
        s.save()
        t.save()

        a=Sponsor(name='hello world')
        a.save()

        rel1=stage_sponsor_yr.create(stage=s,sponsor=a,date=datetime.date(2010,1,1))
        rel1.save()

        rel2=stage_sponsor_yr.create(stage=s,sponsor=a,date=datetime.date(2011,1,1))
        rel2.save()

class MultiRelationTest(TestCase):

    #test accessing child objects
    def test_access_artist_stage0(self):
        s=Stage(location=1)
        s.save()
        for i in range(10):
            a=Artist(name=str(i))
            a.save()
            rel=stage_artist_yr.create(stage=s,artist=a,date=datetime.date(2014,10,10))
            rel.save()

        self.assertEqual(len(s.stage_artist_yr_set.all()),10)

    #test accessing parent object
    def test_access_artist_stage1(self):
        s=Stage(location=1)
        s.save()
        for i in range(10):
            a=Artist(name=str(i))
            a.save()
            rel=stage_artist_yr.create(stage=s,artist=a,date=datetime.date(2014,10,10))
            rel.save()

        for a in s.stage_artist_yr_set.all():
            self.assertEqual(a.stage,s)

    #test accessing child objects (same deal as above but with sponsor)
    def test_access_sponsor_stage2(self):
        s=Stage(location=1)
        s.save()
        for i in range(10):
            a=Sponsor(name=str(i))
            a.save()
            rel=stage_sponsor_yr.create(stage=s,sponsor=a,date=datetime.date(2034,10,12))
            rel.save()

        self.assertEqual(len(s.stage_sponsor_yr_set.all()),10)

    #test accessing parent object
    def test_access_sponsor_stage3(self):
        s=Stage(location=1)
        s.save()
        for i in range(10):
            a=Sponsor(name=str(i))
            a.save()
            rel=stage_sponsor_yr.create(stage=s,sponsor=a,date=datetime.date(2014,10,10))
            rel.save()

        for a in s.stage_sponsor_yr_set.all():
            self.assertEqual(a.stage,s)

class CrossRelationTest(TestCase):
    """
    Tests retrieving members of a time relationship
    """

    #retrieve stage,sponsor,artist
    def test_cross_rel1(self):
        s=Stage(location=10)
        a=Artist(name="Hello World")
        sp=Sponsor(name="Rich People")

        s.save()
        a.save()
        sp.save()

        rel_artist_stage=stage_artist_yr.create(stage=s,artist=a,date=datetime.date(2012,12,25))
        rel_sponsor_stage=stage_sponsor_yr.create(stage=s,sponsor=sp,date=datetime.date(2012,12,25))

        rel_artist_stage.save()
        rel_sponsor_stage.save()

        self.assertEqual(s.stage_artist_yr_set.all()[0],rel_artist_stage)
        self.assertEqual(s.stage_sponsor_yr_set.all()[0],rel_sponsor_stage)

        self.assertEqual(a.stage_artist_yr_set.all()[0],rel_artist_stage)
        self.assertEqual(sp.stage_sponsor_yr_set.all()[0],rel_sponsor_stage)

    #retrieve by relationship
    def test_cross_rel2(self):
        s=Stage(location=10)
        a=Artist(name="Hello World")
        sp=Sponsor(name="Rich People")

        s.save()
        a.save()
        sp.save()

        rel_artist_stage=stage_artist_yr.create(stage=s,artist=a,date=datetime.date(2012,12,25))
        rel_sponsor_stage=stage_sponsor_yr.create(stage=s,sponsor=sp,date=datetime.date(2012,12,25))

        rel_artist_stage.save()
        rel_sponsor_stage.save()

        self.assertEqual(a, rel_artist_stage.artist)
        self.assertEqual(s, rel_artist_stage.stage)

        self.assertEqual(sp, rel_sponsor_stage.sponsor)
        self.assertEqual(s, rel_artist_stage.stage)

class SimulationsTest(TestCase):

    def test_simulation(self):
        #create simulation of the acl festival
        stages=[]
        sponsors=[]
        artists=[]

        #initialize
        for i in range(3):
            x=Stage(location=i)
            y=Artist(name=str(i))
            z=Sponsor(name=str(i))
            x.save()
            y.save()
            z.save()
            stages.append(x)
            artists.append(y)
            sponsors.append(z)

        stage_sponsor=zip(stages,sponsors)
        stage_artist=zip(stages,artists)

        stage_sponsor_rels=list(map(
            lambda elements: stage_sponsor_yr.create(stage=elements[0],sponsor=elements[1],date=datetime.date(2011,3,4)),
            stage_sponsor))
        stage_artist_rels=list(map(
            lambda elements: stage_artist_yr.create(stage=elements[0],artist=elements[1],date=datetime.date(2011,3,4)),
            stage_artist))

        for x,y in zip(stage_sponsor_rels,stage_artist_rels):
            x.save()
            y.save()

        #check that relationships see them
        self.assertEqual(len(list(filter(lambda x: x.stage.location == int(x.sponsor.name), stage_sponsor_rels))), 3)
        self.assertEqual(len(list(filter(lambda x: x.stage.location == int(x.artist.name), stage_artist_rels))), 3)

        self.assertEqual(3, len(stage_artist_yr.objects.all()))
        self.assertEqual(3, len(stage_sponsor_yr.objects.all()))

        #check that they see relationships
        for i in range(3):
            a=Artist.objects.filter(name=str(i))[0]
            for st in a.stage_artist_yr_set.all():
                self.assertEqual(st.stage.location, int(a.name))

            sp=Sponsor.objects.filter(name=str(i))[0]
            for st in sp.stage_sponsor_yr_set.all():
                self.assertEqual(st.stage.location, int(sp.name))

            st=Stage.objects.filter(location=i)[0]
            for ar in st.stage_artist_yr_set.all():
                self.assertEqual(int(ar.artist.name),st.location)

            for s in st.stage_sponsor_yr_set.all():
                self.assertEqual(int(s.sponsor.name),st.location)

        #start adding 2012 info

        #artist1 plays again on stage 1
        m=Artist.objects.filter(name='1')[0]
        sta=Stage.objects.filter(location=1)[0]
        m.stage_artist_yr_set.add(stage_artist_yr.create(stage=sta, artist=m, date=datetime.date(2012,3,4)))

        #but this time with sponsor 2
        n=Sponsor.objects.filter(name='2')[0]
        n.stage_sponsor_yr_set.add(stage_sponsor_yr.create(stage=sta, sponsor=n, date=datetime.date(2012,3,4)))

        #check that relationships is made
        self.assertEqual(len(stage_sponsor_yr.objects.all()),4)

        #artist2 plays on stage 0 with sponsor 1
        m=Artist.objects.filter(name='2')[0]
        sta=Stage.objects.filter(location=0)[0]
        m.stage_artist_yr_set.add(stage_artist_yr.create(stage=sta, artist=m, date=datetime.date(2012,3,4)))
        n=Sponsor.objects.filter(name='1')[0]
        n.stage_sponsor_yr_set.add(stage_sponsor_yr.create(stage=sta, sponsor=n, date=datetime.date(2012,3,4)))






















