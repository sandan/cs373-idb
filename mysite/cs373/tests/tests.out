~/mysite $ python3 manage.py test cs373/tests/
Creating test database for alias 'default'...
..............................
--------------------------------------------------------
Ran 47 tests in 1.920s

OK
Destroying test database for alias 'default'...
